### ISSUE TITLE
Safari .open() DOM IDBDatabase Exception 8

### ISSUE THREAD

**Issue by tothandras** [dfahlander/Dexie.js#120]
I am getting the following error with Safari:

![error](https://www.dropbox.com/s/vyzd1u62rc559vb/Screenshot%202015-08-08%2014.46.28.png?dl=1)

It seems that it occurs on `open()` call:
```coffeescript
db = new window.Dexie('cache')
db.version(1).stores(stores)
console.log(stores)
db.open().catch (e) -> console.error 'db open', e
```

What can cause the problem?

**Comment by dfahlander** [comment]
Does the same code work on chrome /opera on a non IOS platform?

**Comment by tothandras** [comment]
It works with Chrome and Firefox.

**Comment by dfahlander** [comment]
Probably an issue with native indexedDB support on safari. Either narrow it down by reducong the features used in stores and see if you could work around it.  Another option is to include Indexeddbshim before Dexie when on safari or IOS.

**Comment by dfahlander** [comment]
Where you able to work around this using the polyfill or did you find any other way around?
Thanks,
David

**Comment by tothandras** [comment]
Transactions that are using multiple tables were throwing the exceptions.
After I removed those I still have problems reading out data, usually when
it exceeds a certain size.
We might end up using polyfill too.

Andras

**Comment by dfahlander** [comment]
Aha, that's one of the typical issue with Safari - it can only target a single object store at a time.
There's a bunch of issues with Safari implementation mentioned here https://gist.github.com/nolanlawson/08eb857c6b17a30c1b26. A critical one is that primary keys collide between different object stores. Yes, it's true and very sad. Need to use "globally" unique keys to be sure.

**Comment by tothandras** [comment]
I haven't heard about that. That's really bad..

**Comment by lyonlai** [comment]
@tothandras I've just been through this. The root cause is one of the bug: [IndexedDB transaction throw error for multiple object store](https://bugs.webkit.org/show_bug.cgi?id=136937) from the wish list. I'm lucky that I have only 2 object stores to deal with. So my work around to this is to use two separate databases with only 1 object store inside each. Then it will get you through this point.

**Comment by tothandras** [comment]
Exactly, I have commented here about this on 7th of September.

**Comment by dpogue** [comment]
I just ran into this (on one of my databases, but not another), seems to be coming from this line: https://github.com/dfahlander/Dexie.js/blob/master/src/Dexie.js#L549

**Comment by dfahlander** [comment]
They've at least now solved this issue now at Safari in their nightly builds, so things are moving forward. https://bugs.webkit.org/show_bug.cgi?id=136937. Until this fix is released into Safari, either work around this by using only one single store per database, or use IndexedDBShim.

**Comment by dpogue** [comment]
It actually appears (perhaps unintentionally) that if you don't catch the error and abort, later calls to read and write to the database work properly because the rest of Dexie thinks it has been opened. ¯\\\_(ツ)_/¯

**Comment by dfahlander** [comment]
Fixed in 1.3.x.
